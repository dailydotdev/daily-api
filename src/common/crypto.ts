import { webcrypto as crypto } from 'node:crypto';

const bufToBase64 = (x: ArrayBuffer) => Buffer.from(x).toString('base64');

export const encrypt = async (
  input: string,
  key: string,
  algorithmName = 'AES-CBC',
  algorithmLength = 256,
  ivLength = 16,
): Promise<string> => {
  const keyObject = await crypto.subtle.importKey(
    'raw',
    Buffer.from(key, 'utf-8'),
    {
      name: algorithmName,
      length: algorithmLength,
    },
    true,
    ['encrypt', 'decrypt'],
  );
  const iv = crypto.getRandomValues(new Uint8Array(ivLength));
  const encrypted = await crypto.subtle.encrypt(
    {
      name: algorithmName,
      iv,
    },
    keyObject,
    new TextEncoder().encode(input),
  );
  return bufToBase64(iv) + ':' + bufToBase64(encrypted);
};
