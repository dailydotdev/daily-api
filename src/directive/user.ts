import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';

export const deleteUser = async (
  con: DataSource,
  logger: FastifyLoggerInstance,
  userId: string,
  messageId?: string,
) => {
  try {
    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: '404',
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: '404' });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(User).delete(userId);
    });
    if (logger) {
      logger.info(
        {
          userId,
          messageId,
        },
        'deleted user',
      );
    }
  } catch (err) {
    if (logger) {
      logger.error(
        {
          userId,
          messageId,
          err,
        },
        'failed to delete user',
      );
    }
    throw err;
  }
};
