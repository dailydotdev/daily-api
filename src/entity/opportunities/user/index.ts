export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
