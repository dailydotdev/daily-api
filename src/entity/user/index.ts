export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
