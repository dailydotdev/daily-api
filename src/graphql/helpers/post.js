const getFeedParams =
  ({ user, post }, args, rankBy, filters = {}, ignoreUserFilters = false, ip) => {
    const userId = user ? user.userId : null;

    return {
      fields: userId ? [...post.defaultUserFields, 'read'] : post.defaultAnonymousFields,
      filters: Object.assign({}, { before: args.latest }, filters),
      rankBy,
      userId,
      ignoreUserFilters,
      page: args.page,
      pageSize: args.pageSize,
      ip,
    };
  };

const assignType = type => x => Object.assign({}, x, { type });

export default {
  getFeedParams,
  assignType,
};
