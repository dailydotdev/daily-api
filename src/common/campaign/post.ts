import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
