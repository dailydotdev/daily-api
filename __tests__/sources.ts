import {
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
  testMutationErrorCode,
  testQueryErrorCode,
} from './helpers';
import {
  Post,
  SharePost,
  Source,
  SourceFeed,
  SourceMember,
  SourceMemberRoles,
  SquadSource,
  User,
} from '../src/entity';
import { FastifyInstance } from 'fastify';
import request from 'supertest';
import { DataSource } from 'typeorm';
import { randomUUID } from 'crypto';
import createOrGetConnection from '../src/db';
import { usersFixture } from './fixture/user';
import { postsFixture } from './fixture/post';

let app: FastifyInstance;
let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;
let premiumUser: boolean;

const createSource = (id: string, name: string, image: string): Source => {
  const source = new Source();
  source.id = id;
  source.name = name;
  source.image = image;
  source.active = true;
  source.private = false;
  return source;
};

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser, premiumUser),
  );
  client = state.client;
  app = state.app;
});

beforeEach(async () => {
  loggedUser = null;
  premiumUser = false;
  await con
    .getRepository(Source)
    .save([
      createSource('a', 'A', 'http://a.com'),
      createSource('b', 'B', 'http://b.com'),
    ]);
  await saveFixtures(con, User, usersFixture);
  await con.getRepository(SourceMember).save([
    {
      userId: '1',
      sourceId: 'a',
      role: SourceMemberRoles.Owner,
      referralToken: randomUUID(),
      createdAt: new Date(2022, 11, 19),
    },
    {
      userId: '2',
      sourceId: 'a',
      role: SourceMemberRoles.Member,
      referralToken: randomUUID(),
      createdAt: new Date(2022, 11, 20),
    },
    {
      userId: '2',
      sourceId: 'b',
      role: SourceMemberRoles.Owner,
      referralToken: randomUUID(),
      createdAt: new Date(2022, 11, 19),
    },
    {
      userId: '3',
      sourceId: 'b',
      role: SourceMemberRoles.Member,
      referralToken: randomUUID(),
      createdAt: new Date(2022, 11, 20),
    },
  ]);
});

afterAll(() => disposeGraphQLTesting(state));

describe('query sources', () => {
  const QUERY = (first = 10): string => `{
  sources(first: ${first}) {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      node {
        id
        name
        image
        public
      }
    }
  }
}`;

  it('should return only public sources', async () => {
    const res = await client.query(QUERY());
    expect(res.data).toMatchSnapshot();
  });

  it('should flag that more pages available', async () => {
    const res = await client.query(QUERY(1));
    expect(res.data).toMatchSnapshot();
  });

  it('should return only active sources', async () => {
    await con
      .getRepository(Source)
      .save([{ id: 'd', active: false, name: 'D', image: 'http://d.com' }]);
    const res = await client.query(QUERY());
    expect(res.data).toMatchSnapshot();
  });
});

describe('query sourceByFeed', () => {
  const QUERY = `
query SourceByFeed($data: String!) {
  sourceByFeed(feed: $data) {
    id
    name
    image
    public
  }
}`;

  it('should not authorize when not logged in', () =>
    testQueryErrorCode(
      client,
      { query: QUERY, variables: { data: 'https://a.com/feed' } },
      'UNAUTHENTICATED',
    ));

  it('should return null when feed does not exist', async () => {
    loggedUser = '1';
    const res = await client.query(QUERY, {
      variables: { data: 'https://a.com/feed' },
    });
    expect(res.errors).toBeFalsy();
    expect(res.data.sourceByFeed).toEqual(null);
  });

  it('should return the source', async () => {
    loggedUser = '1';
    await con.getRepository(SourceFeed).save({
      feed: 'https://a.com/feed',
      sourceId: 'a',
    });
    const res = await client.query(QUERY, {
      variables: { data: 'https://a.com/feed' },
    });
    expect(res.errors).toBeFalsy();
    expect(res.data.sourceByFeed).toEqual({
      id: 'a',
      name: 'A',
      image: 'http://a.com',
      public: true,
    });
  });
});

describe('query source', () => {
  const QUERY = `
query Source($id: ID!) {
  source(id: $id) {
    id
    name
    image
    public
  }
}
  `;

  it('should not authorize when source does not exist', () =>
    testQueryErrorCode(
      client,
      { query: QUERY, variables: { id: 'notexist' } },
      'NOT_FOUND',
    ));

  it('should not return private source when user is not member', async () => {
    loggedUser = '3';
    await con.getRepository(Source).update({ id: 'a' }, { private: true });
    return testQueryErrorCode(
      client,
      { query: QUERY, variables: { id: 'a' } },
      'FORBIDDEN',
    );
  });

  it('should return source by id', async () => {
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.data).toMatchSnapshot();
  });

  it('should return private source to source members', async () => {
    loggedUser = '1';
    await con.getRepository(Source).update({ id: 'a' }, { private: true });
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.errors).toBeFalsy();
    expect(res.data.source.id).toEqual('a');
  });
});

describe('members field', () => {
  const QUERY = `
query Source($id: ID!) {
  source(id: $id) {
    id
    members {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        node {
          user { id }
          source { id }
          role
        }
      }
    }
  }
}
  `;

  it('should return source members', async () => {
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });

  it('should return source members for private source when the user is a member', async () => {
    loggedUser = '1';
    await con.getRepository(Source).update({ id: 'a' }, { private: true });
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.errors).toBeFalsy();
    expect(res.data.source.members).toMatchSnapshot();
  });
});

describe('query sourceMembers', () => {
  const QUERY = `
query SourceMembers($id: ID!) {
  sourceMembers(sourceId: $id) {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      node {
        user { id }
        source { id }
      }
    }
  }
}
  `;

  it('should not authorize when source does not exist', () =>
    testQueryErrorCode(
      client,
      { query: QUERY, variables: { id: 'notexist' } },
      'NOT_FOUND',
    ));

  it('should return source members of public source', async () => {
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });

  it('should return source members of private source when user is a member', async () => {
    loggedUser = '1';
    await con.getRepository(Source).update({ id: 'a' }, { private: true });
    const res = await client.query(QUERY, { variables: { id: 'a' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });

  it('should not return source members of private source when user is not a member', async () => {
    loggedUser = '3';
    await con.getRepository(Source).update({ id: 'a' }, { private: true });
    return testQueryErrorCode(
      client,
      { query: QUERY, variables: { id: 'a' } },
      'FORBIDDEN',
    );
  });
});

describe('query mySourceMemberships', () => {
  const QUERY = `
query SourceMemberships {
  mySourceMemberships {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      node {
        user { id }
        source { id }
      }
    }
  }
}
  `;

  it('should not authorize when user is not logged in', () =>
    testQueryErrorCode(client, { query: QUERY }, 'UNAUTHENTICATED'));

  it('should return source memberships', async () => {
    loggedUser = '2';
    const res = await client.query(QUERY);
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });
});

describe('compatibility route /publications', () => {
  it('should return only public sources', async () => {
    const res = await request(app.server).get('/v1/publications').expect(200);
    expect(res.body).toMatchSnapshot();
  });
});

describe('mutation createSquad', () => {
  const MUTATION = `
  mutation CreateSquad($name: String!, $handle: String!, $description: String, $postId: ID!, $commentary: String!) {
  createSquad(name: $name, handle: $handle, description: $description, postId: $postId, commentary: $commentary) {
    id
  }
}`;

  const variables = {
    name: 'Squad',
    handle: 'squad',
    postId: 'p1',
    commentary: 'My comment',
  };

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
        variables,
      },
      'UNAUTHENTICATED',
    ));

  it('should create squad', async () => {
    loggedUser = '1';
    await con.getRepository(Post).save(postsFixture[0]);
    const res = await client.mutate(MUTATION, { variables });
    expect(res.errors).toBeFalsy();
    const newId = res.data.createSquad.id;
    const newSource = await con
      .getRepository(SquadSource)
      .findOneBy({ id: newId });
    expect(newSource.name).toEqual('Squad');
    expect(newSource.handle).toEqual('squad');
    expect(newSource.active).toEqual(false);
    const member = await con.getRepository(SourceMember).findOneBy({
      sourceId: newId,
      userId: '1',
    });
    expect(member.role).toEqual(SourceMemberRoles.Owner);
    const post = await con
      .getRepository(SharePost)
      .findOneBy({ sourceId: newId });
    expect(post.authorId).toEqual('1');
    expect(post.sharedPostId).toEqual('p1');
    expect(post.title).toEqual('My comment');
  });

  it('should throw error on duplicate handles', async () => {
    loggedUser = '1';
    await con.getRepository(SquadSource).save({
      id: randomUUID(),
      handle: variables.handle,
      name: 'Dup squad',
      active: false,
    });
    return testMutationErrorCode(
      client,
      { mutation: MUTATION, variables },
      'GRAPHQL_VALIDATION_FAILED',
    );
  });

  it('should throw error when post does not exist', async () => {
    loggedUser = '1';
    return testMutationErrorCode(
      client,
      { mutation: MUTATION, variables },
      'FORBIDDEN',
    );
  });
});
