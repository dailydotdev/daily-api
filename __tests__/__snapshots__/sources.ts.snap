// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`members field should exclude blocked members from result 1`] = `
Object {
  "source": Object {
    "id": "a",
    "members": Object {
      "edges": Array [
        Object {
          "node": Object {
            "role": "admin",
            "source": Object {
              "id": "a",
            },
            "user": Object {
              "id": "1",
            },
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "dGltZToxNjcxNDA4MDAwMDAw",
        "hasNextPage": false,
      },
    },
  },
}
`;

exports[`members field should return source members 1`] = `
Object {
  "source": Object {
    "id": "a",
    "members": Object {
      "edges": Array [
        Object {
          "node": Object {
            "role": "member",
            "source": Object {
              "id": "a",
            },
            "user": Object {
              "id": "2",
            },
          },
        },
        Object {
          "node": Object {
            "role": "admin",
            "source": Object {
              "id": "a",
            },
            "user": Object {
              "id": "1",
            },
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "dGltZToxNjcxNDA4MDAwMDAw",
        "hasNextPage": false,
      },
    },
  },
}
`;

exports[`members field should return source members for private source when the user is a member 1`] = `
Object {
  "edges": Array [
    Object {
      "node": Object {
        "role": "member",
        "source": Object {
          "id": "a",
        },
        "user": Object {
          "id": "2",
        },
      },
    },
    Object {
      "node": Object {
        "role": "admin",
        "source": Object {
          "id": "a",
        },
        "user": Object {
          "id": "1",
        },
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "dGltZToxNjcxNDA4MDAwMDAw",
    "hasNextPage": false,
  },
}
`;

exports[`query source current member should return current member as admin 1`] = `
Object {
  "source": Object {
    "currentMember": Object {
      "permissions": Array [
        "view",
        "post",
        "leave",
        "invite",
        "comment_delete",
        "post_delete",
        "post_pin",
        "member_remove",
        "edit",
        "member_unblock",
        "view_blocked_members",
        "welcome_post_edit",
        "member_role_update",
        "post_limit",
        "delete",
      ],
      "role": "admin",
      "roleRank": 10,
    },
    "id": "a",
  },
}
`;

exports[`query source current member should return current member as blocked 1`] = `
Object {
  "source": Object {
    "currentMember": Object {
      "permissions": Array [],
      "role": "blocked",
      "roleRank": -1,
    },
    "id": "a",
  },
}
`;

exports[`query source current member should return current member as member 1`] = `
Object {
  "source": Object {
    "currentMember": Object {
      "permissions": Array [
        "view",
        "post",
        "leave",
        "invite",
      ],
      "role": "member",
      "roleRank": 0,
    },
    "id": "a",
  },
}
`;

exports[`query source current member should return null for annonymous users 1`] = `
Object {
  "source": Object {
    "currentMember": null,
    "id": "a",
  },
}
`;

exports[`query source current member should return null for user that's not in the source 1`] = `
Object {
  "source": Object {
    "currentMember": null,
    "id": "a",
  },
}
`;

exports[`query source should return source by id 1`] = `
Object {
  "source": Object {
    "id": "a",
    "image": "http://image.com/a",
    "name": "A",
    "public": true,
  },
}
`;

exports[`query sourceMembers should return blocked users only when user is a moderator 1`] = `
Object {
  "sourceMembers": Object {
    "edges": Array [
      Object {
        "node": Object {
          "role": "blocked",
          "roleRank": -1,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "2",
            "name": "Tsahi",
            "username": "tsahidaily",
          },
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "dGltZToxNjcxNDk0NDAwMDAw",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query sourceMembers should return blocked users only when user is the admin 1`] = `
Object {
  "sourceMembers": Object {
    "edges": Array [
      Object {
        "node": Object {
          "role": "blocked",
          "roleRank": -1,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "2",
            "name": "Tsahi",
            "username": "tsahidaily",
          },
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "dGltZToxNjcxNDk0NDAwMDAw",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query sourceMembers should return source members of private source when user is a member 1`] = `
Object {
  "sourceMembers": Object {
    "edges": Array [
      Object {
        "node": Object {
          "role": "admin",
          "roleRank": 10,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "1",
            "name": "Ido",
            "username": "idoshamun",
          },
        },
      },
      Object {
        "node": Object {
          "role": "member",
          "roleRank": 0,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "2",
            "name": "Tsahi",
            "username": "tsahidaily",
          },
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "dGltZToxNjcxNDk0NDAwMDAw",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query sourceMembers should return source members of public source 1`] = `
Object {
  "sourceMembers": Object {
    "edges": Array [
      Object {
        "node": Object {
          "role": "admin",
          "roleRank": 10,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "1",
            "name": "Ido",
            "username": "idoshamun",
          },
        },
      },
      Object {
        "node": Object {
          "role": "member",
          "roleRank": 0,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "2",
            "name": "Tsahi",
            "username": "tsahidaily",
          },
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "dGltZToxNjcxNDk0NDAwMDAw",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query sourceMembers should return source members of public source without blocked members 1`] = `
Object {
  "sourceMembers": Object {
    "edges": Array [
      Object {
        "node": Object {
          "role": "admin",
          "roleRank": 10,
          "source": Object {
            "id": "a",
          },
          "user": Object {
            "id": "1",
            "name": "Ido",
            "username": "idoshamun",
          },
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "dGltZToxNjcxNDA4MDAwMDAw",
      "hasNextPage": false,
    },
  },
}
`;
